rskj-core/src/main/java/co/rsk/core/SenderResolverVisitor.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2019 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.core;
+
+import co.rsk.metrics.profilers.Metric;
+import co.rsk.metrics.profilers.Profiler;
+import co.rsk.metrics.profilers.ProfilerFactory;
+import co.rsk.remasc.RemascTransaction;
+import org.ethereum.core.ImmutableTransaction;
+import org.ethereum.core.Transaction;
+import org.ethereum.crypto.ECKey;
+import org.ethereum.vm.program.InternalTransaction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.SignatureException;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+public class SenderResolverVisitor implements TransactionVisitor<RskAddress> {
+
+    private static final Logger logger = LoggerFactory.getLogger(SenderResolverVisitor.class);
+    private static final Profiler profiler = ProfilerFactory.getInstance();
+
+    private Map<Transaction, RskAddress> senders = new WeakHashMap<>();
+
+    @Override
+    public RskAddress visit(Transaction transaction) {
```

Comment:
I'm worried someone might use this overload with a subclass. We could have an abstract class/interface.

Reply:
Agree, but the real solution involves a redefinition of all this class hierarchy and I think this PR doesn't make that decision any harder for the future